### ISSUE TITLE
VEText Migration - Update Twilio provider logic to use sms sender

### ISSUE THREAD

**Issue by mjones-oddball** [department-of-veterans-affairs/notification-api#747]
Currently our Twilio integration uses parameter store and a single from number instead of using our existing sms sender pattern.

In order to utilize the Twilio Messaging Service as VEText is doing today, we will need to support creating Twilio notification requests with a sender phone number OR message service id. Message service id (if populated) should be the primary choice when the provider is Twilio. Otherwise, use sender phone number.

https://www.twilio.com/docs/messaging/services

- [x] Update the parameter store from number method to be available for testing purposes only (or remove)
- [x] Update the twilio provider method to utilize our sms_sender logic and check for the new message service id value, if found it should use that request method, otherwise it should use the specified phone number method
- [x] Unit tests

## History behind this request
VEText is currently using Twilio and needs to send messages through VA Notify, which uses Amazon. They have millions of lives messages going out each month. Our goal is to minimize the engineering changes needed and impact to Veterans receiving these messages between now and the cutover to Amazon.

We evaluated accepting sender phone number or service id in the SMS request body and ruled this out as an option because it would require changes from both us and VEText when switching from Twilio to Amazon. Because of this we agreed that VEText would use the sms_sender_id in the SMS request body and VA Notify would support one additional field in our SMS Sender database table to capture Twilio's service id value. Why is this used? In Twilio you can set up features if using a service id. E.g. fallback from short code to a long code. This is used on the highest volume use cases (short codes) that VEText has. They also have use cases where a specific long code is assigned to a Veteran who can't receive messages from short codes. In that case we send the Twilio message from a specific long code.

What this means:
VEText sends an SMS notification request including: 

- Template id
- Personalisation
- Recipient id or phone number
- sms_sender_id

VA Notify will create sms senders for each use case they have and provide the sms_sender_ids to them. 
VA Notify will include both the sms_sender AND twilio service id (sms_sender_specifics) for those use cases, so that after the cutover to Amazon, the correct value is used.

The Twilio provider logic should A) use our same sms sender model that we use for Amazon B) should check for sms_sender_specifics first and use that in the request to Twilio if found, else use the sms_sender

Amazon should not use the sms_sender_specifics field

**Comment by mjones-oddball** [comment]
@Jake-Uhteg added QA tag

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/747?planning-poker) with ZenHub @EvanParish @ianperera @jessecanderson @k-macmillan @kalbfled @maham360 @jakehova

**Comment by kalbfled** [comment]
@mjones-oddball @jakehova What is the format of a message service id?

**Comment by kalbfled** [comment]
@mjones-oddball @zlaw I just created a PR for this issue, and I expect to have to make revisions.  I propose to re-point this task as an 8.

**Comment by kalbfled** [comment]
@mjones-oddball PR 798 implements this feature, but it doesn't use the sms_sender_specifics field added for #748.  As written in PR 798, the request to /v2/notifications/sms can contain a message_service_sid attribute.  If it does, and if Twilio is used as the provider, the message_service_sid will by used instead of a sender phone number.

I realized afterwards that the intend probably is to look-up the message_service_sid in the sms_sender_specifics attribute, but the way the code is structured makes me think that this apparently simple change will require a significant amount of refactoring.  I propose to do that under a new ticket.  This one already contains changes to 21 files.

**Comment by mjones-oddball** [comment]
@kalbfled Yes the intention on this ticket was to update the Twilio code to use the existing sms sender model that our Amazon logic uses today instead of parameter store. We would then update it (for Twilio) to look up the message service sid in the new field that was added (sms_sender_specifics). If found, use that instead of sms_sender (aka the from phone number). 

The why behind all of this is that this is only phase 1 of the migration. VEText needs to make code updates to send SMS requests to us instead of directly to Twilio. They want to minimize changes needed when we're ready to switch to Amazon. So if they send us the SMS request per usual with sms_sender_id, then the logic of what number to send from would all be handled behind the scenes without additional engineering work. So that when we update the provider, it will naturally go back to using the sms_sender (from phone number) instead of the Twilio specific service id value. We'll use refinement to talk through where we're at and next steps.
CC @k-macmillan @zlaw

**Comment by kalbfled** [comment]
@mjones-oddball @k-macmillan I have revised my PR to utilize *sms_sender_specifics* and not accept *message_service_sid* in the POST request.  The PR is back in draft status while @EvanParish adds new unit tests.

**Comment by EvanParish** [comment]
I've taken over the work since Dave has several other tasks going on and Kyle identified a way to make this change more generic and reduce future tech debt. I have added `**kwargs` to the Twilio send_sms method, but with the current changes I'm still getting some failing unit tests.

**Comment by EvanParish** [comment]
I have added some unit testing to make sure the new path is being taken when trying to send a message to a service that has sms_sender_specifics information.

**Comment by EvanParish** [comment]
I ended up running into some issues and misinterpreted the purpose of the rework. This is going to require some more work in order to smooth out the problems. I'm going to be trying to move the twilio-specific code into the twilio provider, but I will need to figure out how to handle an issue I'm getting with having a circular dependency.

**Comment by EvanParish** [comment]
I believe I have fixed the issues with the rework. I also added unit testing for the new path taken when Twilio uses the `sms_sender_specifics` field. Logging has been added so we can track which path the message takes in the code to make sure the new code is working as expected.

**Comment by EvanParish** [comment]
I'm still having some trouble getting the Twilio provider to send messages through in Dev. I also needed to update the template code in `rest.py` to get it to update the `reply_to_text` field.

**Comment by EvanParish** [comment]
Working with Kyle, we discovered that when it selects the `sms_sender`, it was just grabbing the first one available on the service with the specified number. I had to update it to consider the `sms_sender`'s template id if it was available. This still needs testing in Dev, but the unit tests are now passing and it should be grabbing the correct `sms_sender`.

**Comment by EvanParish** [comment]
I have had successful tests in the Dev environment. The PR has to be recreated with a new branch since I had merged master into the existing one, adding many unrelated changes to the branch.

**Comment by EvanParish** [comment]
The PR has been updated with details about the setup and testing for this change and everything seems to be working. I've made changes according to other's suggestions and these changes should be ready to merge soon. I also added a flowchart for the notification path through the code so it's hopefully easier to follow in the future. 
https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Support/sms_notifications_flowchart.md
